main.go

```diff
@@ -80,94 +143,204 @@ func main() {
 		TLSNextProto: make(map[string]func(*http.Server, *tls.Conn, http.Handler), 0),
 	}
 
-	fmt.Printf("Login endpoint available at: http://%v/login\nCallback endpoint available at: %v\n\n", server.Addr, acpOAuthConfig.RedirectURL)
+	fmt.Printf("Login endpoint available at: http://localhost:%v/login\nCallback endpoint available at: %v\n\n", config.PORT, client.Config.RedirectURL)
 	if err := server.ListenAndServe(); err != http.ErrServerClosed {
 		log.Fatalln(err)
 	} else {
 		log.Println("server closed!")
 	}
 }
 
-func login(writer http.ResponseWriter, request *http.Request) {
-	var challenge string
+func login(authorizeURL string) func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		http.Redirect(w, r, authorizeURL, http.StatusTemporaryRedirect)
+	}
+}
 
-	//If PKCE is enabled, generate code verifier and challenge.
-	if *pkceEnabled {
+func callback(client acp.Client, csrf acp.CSRF) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
 		var (
-			encodedVerifier    string
-			encodedCookieValue string
-			err                error
+			code = r.URL.Query().Get("code")
+			err  error
 		)
 
-		verifier := make([]byte, challengeLength)
-		if _, err = io.ReadFull(rand.Reader, verifier); err != nil {
-			log.Printf("error while generating challenge, %v\n", err)
+		if token, err = client.Exchange(code, csrf.State, csrf); err != nil {
+			log.Printf("%v\n", err)
+			w.Write([]byte(err.Error()))
```

Comment:
If you already have an error template, you could render it here with some user-friendly message.
There are some other places (especially in web routes) where you could render an error page as well.